core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -207,4 +289,130 @@ public void doXContent(XContentBuilder builder, Params params) throws IOExceptio
+        List<Object> terms = new ArrayList<>();
+        if (isTermsLookupQuery()) {
+            String lookupIndex = (this.lookupIndex == null) ? context.index().name() : this.lookupIndex;
+            handleTermsLookup(new TermsLookup(lookupIndex, lookupType, lookupId, lookupRouting, lookupPath, context.parseContext()), context.client(), terms);
+        } else {
+            terms.addAll(values);
+        }
+        if (terms.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+        return handleTermsQuery(terms, fieldName, context, disableCoord);
+    }
+
+    static void handleTermsLookup(TermsLookup lookup, Client client, List<Object> terms) {
+        GetRequest getRequest = new GetRequest(lookup.getIndex(), lookup.getType(), lookup.getId()).preference("_local").routing(lookup.getRouting());
+        getRequest.copyContextAndHeadersFrom(SearchContext.current());
+        final GetResponse getResponse = client.get(getRequest).actionGet();
+        if (getResponse.isExists()) {
+            List<Object> extractedValues = XContentMapValues.extractRawValues(lookup.getPath(), getResponse.getSourceAsMap());
+            terms.addAll(extractedValues);
+        }
+    }
+
+    static Query handleTermsQuery(List<Object> terms, String fieldName, QueryShardContext context, boolean disableCoord) {
+        MappedFieldType fieldType = context.fieldMapper(fieldName);
+        String indexFieldName;
+        if (fieldType != null) {
+            indexFieldName = fieldType.names().indexName();
+        } else {
+            indexFieldName = fieldName;
+        }
+
+        Query query;
+        if (context.isFilter()) {  // NO COMMIT: not exactly sure how to handle this case, right now using a filter flag
+            if (fieldType != null) {
+                query = fieldType.termsQuery(terms, context);
+            } else {
+                BytesRef[] filterValues = new BytesRef[terms.size()];
+                for (int i = 0; i < filterValues.length; i++) {
+                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
+                }
+                query = new TermsQuery(indexFieldName, filterValues);
+            }
+        } else {
+            BooleanQuery bq = new BooleanQuery(disableCoord);
+            for (Object term : terms) {
+                if (fieldType != null) {
+                    bq.add(fieldType.termQuery(term, context), BooleanClause.Occur.SHOULD);
+                } else {
+                    bq.add(new TermQuery(new Term(indexFieldName, BytesRefs.toBytesRef(term))), BooleanClause.Occur.SHOULD);
+                }
+            }
+            query = bq;
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException validationException = null;
+        if (this.fieldName == null) {
+            validationException = addValidationError("field name cannot be null.", validationException);
+        }
```

Comment:
ok, I think we may have to make this more consistent. As I previously suggested let's introduce an inner terms lookup builder object. I think when TermsQueyBuilder doesn't hold a terms lookup though the values need to be non null so we should validate it?